pkg/logservice/snapshot.go

```diff
@@ -0,0 +1,357 @@
+// Copyright 2021 - 2022 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logservice
+
+import (
+	"fmt"
+	"path/filepath"
+	"sort"
+	"sync"
+
+	"github.com/cockroachdb/errors"
+	"github.com/lni/vfs"
+)
+
+const (
+	defaultExportedDirMode = 0755
+	snapshotPathPattern    = "snapshot-%016X"
+)
+
+// ISnapshotManager is an interface that managers snapshots.
+type ISnapshotManager interface {
+	// Init initialize snapshots by loading exported snapshots.
+	Init(shardID uint64, replicaID uint64) error
+	// Count returns the number of snapshots in the manager.
+	Count(shardID uint64, replicaID uint64) int
+	// Add adds a new snapshot for specified shard.
+	Add(shardID uint64, replicaID uint64, index uint64) error
+	// Remove removes snapshots whose index is LE than index.
+	Remove(shardID uint64, replicaID uint64, index uint64) error
+	// EvalImportSnapshot returns the source directory and index of
+	// the biggest snapshot of the shard.
```

Comment:
What is "biggest"?

Reply:
there are several snapshots which are not visible to dragonboat in the export directory. their indexes are sorted. when some of them are caught up by truncated lsn, the index-biggest one will be chosen to import into dragonboat and make it visible.